source/opt/aggressive_dead_code_elim_pass.cpp

```diff
@@ -123,40 +129,193 @@ void AggressiveDCEPass::ProcessLoad(uint32_t varId) {
   live_local_vars_.insert(varId);
 }
 
+uint32_t AggressiveDCEPass::MergeBlockIdIfAny(
+    const ir::BasicBlock& blk, uint32_t* cbid) const {
+  auto merge_ii = blk.cend();
+  --merge_ii;
+  uint32_t mbid = 0;
+  *cbid = 0;
+  if (merge_ii != blk.cbegin()) {
+    --merge_ii;
+    if (merge_ii->opcode() == SpvOpLoopMerge) {
+      mbid = merge_ii->GetSingleWordInOperand(kLoopMergeMergeBlockIdInIdx);
+      *cbid = merge_ii->GetSingleWordInOperand(kLoopMergeContinueBlockIdInIdx);
+    }
+    else if (merge_ii->opcode() == SpvOpSelectionMerge) {
+      mbid = merge_ii->GetSingleWordInOperand(
+          kSelectionMergeMergeBlockIdInIdx);
+    }
+  }
+  return mbid;
+}
+
+void AggressiveDCEPass::ComputeStructuredSuccessors(ir::Function* func) {
+  // If header, make merge block first successor. If a loop header, make
+  // the second successor the continue target.
+  for (auto& blk : *func) {
+    uint32_t cbid;
+    uint32_t mbid = MergeBlockIdIfAny(blk, &cbid);
+    if (mbid != 0) {
+      block2structured_succs_[&blk].push_back(id2block_[mbid]);
+      if (cbid != 0)
+        block2structured_succs_[&blk].push_back(id2block_[cbid]);
+    }
+    // add true successors
+    blk.ForEachSuccessorLabel([&blk, this](uint32_t sbid) {
+      block2structured_succs_[&blk].push_back(id2block_[sbid]);
+    });
+  }
+}
+
+void AggressiveDCEPass::ComputeStructuredOrder(
+    ir::Function* func, std::list<ir::BasicBlock*>* order) {
+  // Compute structured successors and do DFS
+  ComputeStructuredSuccessors(func);
+  auto ignore_block = [](cbb_ptr) {};
+  auto ignore_edge = [](cbb_ptr, cbb_ptr) {};
+  auto get_structured_successors = [this](const ir::BasicBlock* block) {
+      return &(block2structured_succs_[block]); };
+  // TODO(greg-lunarg): Get rid of const_cast by making moving const
+  // out of the cfa.h prototypes and into the invoking code.
+  auto post_order = [&](cbb_ptr b) {
+      order->push_front(const_cast<ir::BasicBlock*>(b)); };
+
+  spvtools::CFA<ir::BasicBlock>::DepthFirstTraversal(
+      &*func->begin(), get_structured_successors, ignore_block, post_order,
+      ignore_edge);
+}
+
+bool AggressiveDCEPass::IsStructuredIfHeader(ir::BasicBlock* bp,
+      ir::Instruction** mergeInst, ir::Instruction** branchInst,
+      uint32_t* mergeBlockId) {
+  auto ii = bp->end();
+  --ii;
+  if (ii->opcode() != SpvOpBranchConditional)
+    return false;
+  if (ii == bp->begin())
+    return false;
+  *branchInst = &*ii;
+  --ii;
+  if (ii->opcode() != SpvOpSelectionMerge)
+    return false;
+  *mergeInst = &*ii;
+  *mergeBlockId = ii->GetSingleWordInOperand(kSelectionMergeMergeBlockIdInIdx);
+  return true;
+}
+
+void AggressiveDCEPass::ComputeBlock2BranchMaps(
```

Comment:
This reminds me of the control dependence graph (CDG).  If we keep the code this way, it might be a good idea to move this analysis out into it own things.  Build the CDG, and then it can easily be reused by other transformations that need it.  It is not needed by many thing, but at least is will be available.  Plus it will make it easier to test on it own.

Reply:
I think we should try to be careful about what we assume.  If we are doing something that only works because we have structured control flow, we should add asserts and comments indicating that.  They should be close to the part of the code that uses that assumption.  We don't want the OpenCL world accidentally use this code.  It also makes it obvious which parts need to be updated if we want this optimization to work for things other than Vulkan shaders.

Also because the concept of control dependence is generally known, it might be a good idea to talk about it in those terms.  When I see a function name ComputeBlock2BranchMaps, I have no idea how they are related.  Are they the branches at the end of said block?  Even if there are comments in the header, I do not want to go to the header every time I see a function call.

Reply:
> I think we should try to be careful about what we assume. If we are doing something that only works because we have structured control flow, we should add asserts and comments indicating that.

We've been handling this with the first guard (and early return) at the top of ProcessImpl.  It checks for Shader, which then requires structured control flow.  The assumption that structured control flow is required is somewhat pervasive, and I wouldn't want to pepper all the existing code with repeat of the same assert. It's something that all of us working in this code are now accustomed to.  

Perhaps it merits a comment, but I think the fact that we're using a structuredOrder parameter is a very very strong signal.  Also, the code is peppered with ....MergBlockId...  which again is a very strong signal.

Given all that I'm inclined to say we don't need yet another signal boost that this requires/assumes structured control flow.

Reply:
> Even if there are comments in the header, I do not want to go to the header every time I see a function call.

I've given pretty strong guidance to follow the Google C++ style guide on this topic.  See 
https://google.github.io/styleguide/cppguide.html#Function_Comments

In particular:
* The method comment is given in the header.  Don't repeat it at the function body because it's liable to get out of date.
* The method comment describes what the function does from the client's perspective
* Comments about the implementation details (if it's non-obvious) are placed inside the function body.

How does this apply in this instance?   The PR is good, in my view, regarding placement of the comments.  I think it could use more explanatory text about the relation of the two block2merge_ and block2branch_ to the more general CDG concept.  I would put that in one place, probably at the declaration of those members.  Another reasonable spot is an explanatory paragraph before the declaration of this method.

Reply:
Saying it another way, I'm strongly inclined to document the *data*, and then the methods are generally self-explanatory.  (I think Fred Brooks said something about this... though I think it was framed in terms of "tables")

Reply:
Sorry, my suggestion was not about the comments, but the name of the function and data.

Reply:
I agree the names are too terse. I will change the name of this method to ComputeBlock2HeaderMaps and the name of the structures to block2headerMerge and block2headerBranch.